crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -124,10 +128,18 @@ impl<'a> SemanticIndexBuilder<'a> {
         &mut self.ast_ids[scope_id]
     }
 
-    fn add_or_update_symbol(&mut self, name: Name, flags: SymbolFlags) -> ScopedSymbolId {
-        let symbol_table = self.current_symbol_table();
+    fn add_or_update_symbol(&mut self, name: Name, flags: SymbolFlags) -> FileSymbolId {
+        for scope in self.scope_stack.iter().rev().skip(1) {
+            let builder = &self.symbol_tables[*scope];
+
+            if let Some(symbol) = builder.symbol_by_name(&name) {
+                return FileSymbolId::new(*scope, symbol);
+            }
+        }
 
-        symbol_table.add_or_update_symbol(name, flags, None)
+        let scope = self.current_scope();
+        let symbol_table = self.current_symbol_table();
```

Comment:
I considered filtering in the type inference but having types in the symbol table that come from the parent scope feels like a huge foot gun because we would need to filter them whenever we lookup a symbol by name.

Reply:
I don't understand. We don't have types in the symbol table at all. I'm not clear where the foot gun or filtering worry is. 

I explained above why we must have symbols in the symbol table if used in that scope, whether defined in that scope or not.

If you want to leave this PR this way that's ok, I'll just have to revert this change again in the PR that adds control flow.

Reply:
I considered filtering in the type inference but having types in the symbol table that come from the parent scope feels like a huge foot gun because we would need to filter them whenever we lookup a symbol by name.

Reply:
What I'm concerned about is that all call-sites of `symbol_table.symbol_by_name` will have to check whether the returned symbol is actually defined in that scope. I find it rather unexpected to find symbols in a scope that are inherited from the parent and it seems like an easy source for bugs. 

I'm okay patching this up in the call site, but how do we avoid that we don't have to do this in multiple places. For example, how can we avoid that we don't need to do it in the `SemanticNode::symbol` implementation of `ExprName` and in the type inference code? I guess one solution is to just implement it on `ExprName`, but it is then important that we only use that method too perform the lookup making it even more important that we don't expose `symbol_table` directly.

Reply:
The symbol in the inner scope is a separate symbol from the symbol in the outer scope. In some cases one of a symbol's definitions may refer to the corresponding symbol in the outer scope, but this will be correctly reflected in the definitions; we'll have a `Definition:Nonlocal` or similar at the start of control flow of the inner scope.

This precisely maps to how free variables are handled in the runtime: the two scopes have separate symbols of the same name, with separate local variable storage in the frame for each scope. At the start of the inner frame the local variable storage for a free variable is populated with a "cell" containing the value from the outer scope.

So I don't think it's correct to assume that callers even _should_ do the filtering you're describing, or that it will be a bug if they don't do it; it depends entirely on why the caller is asking for a symbol. Symbols exist in a scope, and if the caller asks for the symbol in scope X, they should get that symbol, not the one in some other outer scope. If the caller then wants to know where the symbol's value might have been defined, they should use Definitions to discover that, and Definitions will give them the right answer, possibly via `Definition::Nonlocal`. (But in most/all cases, the caller shouldn't be worrying about this at all, they should just ask for the type of the symbol, and type inference will already have handled the lookup for them.)

Even in type inference name lookup, the filtering you describe is only a temporary hack until we add control flow. Once we have control flow, even type inference won't do the filtering (or the walking up scopes) that you currently do when inferring the type of an `ExprName`; it will just look for the reachable definition(s) in the local scope, one of which might be `Definition::Nonlocal`, and resolve the types of those definitions appropriately.

If you have a specific use case for `Node::symbol` in mind that you are worried might do the wrong thing for free variables, we can discuss how to avoid that.

Reply:
Another way of framing this: Python's name lookup rules are inherently quite complex (e.g. with `nonlocal`, `global`, the special rules about class scopes and annotation scopes, etc). Regardless of how we handle this particular case of free variables, we can't get rid of this complexity.

One way of handling this, which is the direction you're headed in with this change, is to embed all of these name lookup rules inside how we build the symbol table itself. But this still means we have to duplicate this logic both in the symbol table builder and in every user of it. In this PR, the scope-walking is now duplicated both here and in type inference, and it has the same bugs regarding class and annotation scopes in both places, that would now have to be fixed separately in both places. So if we have multiple callers needing to do name lookup, this approach doesn't avoid duplication of the logic, it just adds one additional place (the symbol table builder) where the same logic has to exist.

This approach also has trouble with `nonlocal` or `global` symbols that are defined both in the inner and outer scope; we can't treat these as just a single symbol, without ignoring the definition(s) in the inner scope, or else having outer scopes depend on inner ones, and if we treat them as two separate symbols, then we are right back to needing `Definition::Nonlocal` anyway.

Reply:
I changed it. I'm still hesitant about creating symbol table entries for symbols that don't exist in that scope (and I think that's different from `nonlocal` because we would create a local symbol in that case). But we can discuss it in detail when you implement `nonlocal`.

Reply:
Another example where I think we need the filtering. `public_symbol` 

https://github.com/astral-sh/ruff/blob/085c33d48914db75e9010d86f03ffdabf2898151/crates/red_knot_python_semantic/src/semantic_index.rs#L61-L70

It currently returns `Some` when looking up `x` in

```python
print(x)
```

Reply:
There's no need for filtering there. It's fine if the symbol returns Some, and its only definition in that scope is unbound. We have to handle the possibility of unbound regardless, so this doesn't introduce any new problem or case to handle.

In the public symbols case it would be fine either way; in most typical lookups of names used within the scope, it will be simpler if we only have to handle a definition of Unbound or Nonlocal, rather than having to handle both that case and "symbol doesn't exist" (in the same way.)